pkg/util/active_groups.go

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package util
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/grafana/dskit/services"
+	"go.uber.org/atomic"
+)
+
+type ActiveGroups struct {
+	mu                sync.RWMutex
+	timestampsPerUser map[string]map[string]*atomic.Int64 // map[user][group] -> timestamp
+	maxGroupsPerUser  int
+}
+
+func NewActiveGroups(maxGroupsPerUser int) *ActiveGroups {
+	return &ActiveGroups{
+		timestampsPerUser: map[string]map[string]*atomic.Int64{},
+		maxGroupsPerUser:  maxGroupsPerUser,
+	}
+}
+
+// UpdateGroupTimestampForUser function is only guaranteed to update to timestamp
+// provided even if it is smaller than the existing value
+func (ag *ActiveGroups) UpdateGroupTimestampForUser(userID, group string, now time.Time) {
+	ts := now.UnixNano()
+	ag.mu.RLock()
+	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
+		ag.mu.RUnlock()
+		groupTs.Store(ts)
+		return
+	}
+	ag.mu.RUnlock()
+
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	if groupTimestamps == nil {
+		ag.timestampsPerUser[userID] = map[string]*atomic.Int64{group: atomic.NewInt64(ts)}
+		return
+	}
+
+	groupTs := ag.timestampsPerUser[userID][group]
+
+	if groupTs != nil {
+		groupTs.Store(ts)
+		return
+	}
```

Comment:
Ignorable style nit, this could be clearer:

```suggestion
	if groupTs := ag.timestampsPerUser[userID][group]; groupTs != nil {
		groupTs.Store(ts)
		return
	}
```